malformed pipeline manifest causes a pachd panic

```
2018-01-30T00:21:17Z INFO pps.API.CreatePipeline {"request":{"pipeline":{"name":"user_invalid_submission2LIC2_inference"},"description":"Given any docker image, runs a common entry point for a user's inference model"}} 
2018-01-30T00:21:17Z INFO pps.API.call64 {"duration":0.000050014,"request":{"pipeline":{"name":"user_invalid_submission2LIC2_inference"},"description":"Given any docker image, runs a common entry point for a user's inference model"},"response":null} 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x173aa5d]

goroutine 15241 [running]:
github.com/pachyderm/pachyderm/src/server/pps/server.setPipelineDefaults(0xc4205a4500)
	/go/src/github.com/pachyderm/pachyderm/src/server/pps/server/api_server.go:1569 +0x3d
github.com/pachyderm/pachyderm/src/server/pps/server.(*apiServer).CreatePipeline(0xc4201a0480, 0x7fd3d5300220, 0xc4203c0480, 0xc4208be210, 0x0, 0x0, 0x0)
	/go/src/github.com/pachyderm/pachyderm/src/server/pps/server/api_server.go:1390 +0x4bd
github.com/pachyderm/pachyderm/src/server/vendor/github.com/pachyderm/pachyderm/src/client/pps._API_CreatePipeline_Handler(0x1bd04e0, 0xc4201a0480, 0x7fd3d5300220, 0xc4203c0480, 0xc420232370, 0x0, 0x0, 0x0, 0xc42088bac0, 0xc42088bac8)
	/go/src/github.com/pachyderm/pachyderm/src/server/vendor/github.com/pachyderm/pachyderm/src/client/pps/pps.pb.go:3143 +0x276
github.com/pachyderm/pachyderm/src/server/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc4200cac80, 0x2a18c60, 0xc4205498c0, 0xc4202c0300, 0xc420348ab0, 0x2a91d40, 0x0, 0x0, 0x0)
	/go/src/github.com/pachyderm/pachyderm/src/server/vendor/google.golang.org/grpc/server.go:806 +0xab4
github.com/pachyderm/pachyderm/src/server/vendor/google.golang.org/grpc.(*Server).handleStream(0xc4200cac80, 0x2a18c60, 0xc4205498c0, 0xc4202c0300, 0x0)
	/go/src/github.com/pachyderm/pachyderm/src/server/vendor/google.golang.org/grpc/server.go:1003 +0x1528
github.com/pachyderm/pachyderm/src/server/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc42055e680, 0xc4200cac80, 0x2a18c60, 0xc4205498c0, 0xc4202c0300)
	/go/src/github.com/pachyderm/pachyderm/src/server/vendor/google.golang.org/grpc/server.go:552 +0x9f
created by github.com/pachyderm/pachyderm/src/server/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/github.com/pachyderm/pachyderm/src/server/vendor/google.golang.org/grpc/server.go:550 +0xa1
```

--- a/src/client/pkg/errors/errors.go
+++ b/src/client/pkg/errors/errors.go
@@ -33,6 +33,7 @@ var (
 	WithStack = errors.WithStack
 )
 
+// StackTrace is stack of Frames from innermost (newest) to outermost (oldest).
 type StackTrace = errors.StackTrace
 
 // EnsureStack will add a stack onto the given error only if it does not already

--- a/src/client/pkg/errors/errors.go
+++ b/src/client/pkg/errors/errors.go
@@ -33,6 +33,8 @@ var (
 	WithStack = errors.WithStack
 )
 
+type StackTrace = errors.StackTrace
+
 // EnsureStack will add a stack onto the given error only if it does not already
 // have a stack. If err is nil, EnsureStack returns nil.
 func EnsureStack(err error) error {
@@ -48,7 +50,7 @@ func EnsureStack(err error) error {
 }
 
 // Frame is the type of a StackFrame, it is an alias for errors.Frame.
-type Frame errors.Frame
+type Frame struct{ errors.Frame }
 
 // Callers returns an errors.StackTrace for the place at which it's called.
 func Callers() errors.StackTrace {
@@ -83,7 +85,7 @@ func ForEachStackFrame(err error, f func(Frame)) {
 	}
 	if len(st) > 0 {
 		for _, frame := range st {
-			f(Frame(frame))
+			f(Frame{frame})
 		}
 	}
 }

--- a/src/client/pps/util.go
+++ b/src/client/pps/util.go
@@ -117,15 +117,14 @@ func ValidateGitCloneURL(url string) error {
 	//     ssh_url: "git@github.com:sjezewski/testgithook.git",
 	//     clone_url: "https://github.com/sjezewski/testgithook.git",
 	//     svn_url: "https://github.com/sjezewski/testgithook",
-	invalidErr := errors.Errorf("clone URL is missing .git suffix (example clone URL %v)", exampleURL)
 
 	if !strings.HasSuffix(url, ".git") {
 		// svn_url case
-		return invalidErr
+		return errors.Errorf("clone URL is missing .git suffix (example clone URL %v)", exampleURL)
 	}
 	if !strings.HasPrefix(url, "https://") {
 		// git_url or ssh_url cases
-		return invalidErr
+		return errors.Errorf("clone URL must use https protocol (example clone URL %v)", exampleURL)
 	}
 
 	return nil

--- a/src/server/pachyderm_test.go
+++ b/src/server/pachyderm_test.go
@@ -11558,7 +11558,8 @@ func TestMalformedPipeline(t *testing.T) {
 
 	pipelineName := tu.UniqueString("MalformedPipeline")
 
-	_, err := c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{})
+	var err error
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{})
 	require.YesError(t, err)
 	require.Matches(t, "invalid pipeline spec", err.Error())
 
@@ -11672,6 +11673,14 @@ func TestMalformedPipeline(t *testing.T) {
 		Input:     &pps.Input{Cron: &pps.CronInput{}},
 	})
 	require.YesError(t, err)
+	require.Matches(t, "input must specify a name", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{Cron: &pps.CronInput{Name: "cron"}},
+	})
+	require.YesError(t, err)
 	require.Matches(t, "Empty spec string", err.Error())
 
 	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
@@ -11680,7 +11689,23 @@ func TestMalformedPipeline(t *testing.T) {
 		Input:     &pps.Input{Git: &pps.GitInput{}},
 	})
 	require.YesError(t, err)
-	require.Matches(t, "clone URL is missing", err.Error())
+	require.Matches(t, "clone URL is missing \\(", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{Git: &pps.GitInput{URL: "foobar"}},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "clone URL is missing .git suffix", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{Git: &pps.GitInput{URL: "foobar.git"}},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "clone URL must use https protocol", err.Error())
 
 	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
 		Pipeline:  client.NewPipeline(pipelineName),
@@ -11705,22 +11730,6 @@ func TestMalformedPipeline(t *testing.T) {
 	})
 	require.YesError(t, err)
 	require.Matches(t, "no input set", err.Error())
-}
-
-func TestSimpleCron(t *testing.T) {
-	c := tu.GetPachClient(t)
-	require.NoError(t, c.DeleteAll())
-
-	pipelineName := tu.UniqueString("MalformedPipeline")
-
-	_, err := c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
-		Pipeline:  client.NewPipeline(pipelineName),
-		Transform: &pps.Transform{Cmd: []string{"bash"}, Stdin: []string{"echo foo > /pfs/out/file"}},
-		Input:     &pps.Input{Cron: &pps.CronInput{Spec: "@every 10s"}},
-	})
-	require.NoError(t, err)
-
-	// TODO: the worker fails to symlink active data directory - because there is no name?
 }
 
 func getObjectCountForRepo(t testing.TB, c *client.APIClient, repo string) int {

--- a/src/server/pachyderm_test.go
+++ b/src/server/pachyderm_test.go
@@ -11551,6 +11551,178 @@ func TestKeepRepo(t *testing.T) {
 	require.NoError(t, c.DeletePipeline(pipeline, false))
 }
 
+// Regression test to make sure that pipeline creation doesn't crash pachd due to missing fields
+func TestMalformedPipeline(t *testing.T) {
+	c := tu.GetPachClient(t)
+	require.NoError(t, c.DeleteAll())
+
+	pipelineName := tu.UniqueString("MalformedPipeline")
+
+	_, err := c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{})
+	require.YesError(t, err)
+	require.Matches(t, "invalid pipeline spec", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline: client.NewPipeline(pipelineName)},
+	)
+	require.YesError(t, err)
+	require.Matches(t, "must specify a transform", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "no input set", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:        client.NewPipeline(pipelineName),
+		Transform:       &pps.Transform{},
+		Service:         &pps.Service{},
+		ParallelismSpec: &pps.ParallelismSpec{},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "services can only be run with a constant parallelism of 1", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:     client.NewPipeline(pipelineName),
+		Transform:    &pps.Transform{},
+		HashtreeSpec: &pps.HashtreeSpec{},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "HashtreeSpec.Constant must be > 0", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:   client.NewPipeline(pipelineName),
+		Transform:  &pps.Transform{},
+		SpecCommit: &pfs.Commit{},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "cannot resolve commit with no repo", err.Error())
+
+	dataRepo := tu.UniqueString("TestMalformedPipeline_data")
+	require.NoError(t, c.CreateRepo(dataRepo))
+
+	_, err = c.PutFile(dataRepo, "master", "file", strings.NewReader("foo"))
+	require.NoError(t, err)
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{Pfs: &pps.PFSInput{}},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "input must specify a name", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{Pfs: &pps.PFSInput{Name: "data"}},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "input must specify a repo", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{Pfs: &pps.PFSInput{Repo: dataRepo}},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "input must specify a glob", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     client.NewPFSInput("out", "/*"),
+	})
+	require.YesError(t, err)
+	require.Matches(t, "input cannot be named \"out\"", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{Pfs: &pps.PFSInput{Name: "out", Repo: dataRepo, Glob: "/*"}},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "input cannot be named \"out\"", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{Pfs: &pps.PFSInput{Name: "data", Repo: "dne", Glob: "/*"}},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "dne not found", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input: client.NewCrossInput(
+			client.NewPFSInput("foo", "/*"),
+			client.NewPFSInput("foo", "/*"),
+		),
+	})
+	require.YesError(t, err)
+	require.Matches(t, "name \"foo\" was used more than once", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{Cron: &pps.CronInput{}},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "Empty spec string", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{Git: &pps.GitInput{}},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "clone URL is missing", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{Cross: []*pps.Input{}},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "no input set", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{Union: []*pps.Input{}},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "no input set", err.Error())
+
+	_, err = c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{},
+		Input:     &pps.Input{Join: []*pps.Input{}},
+	})
+	require.YesError(t, err)
+	require.Matches(t, "no input set", err.Error())
+}
+
+func TestSimpleCron(t *testing.T) {
+	c := tu.GetPachClient(t)
+	require.NoError(t, c.DeleteAll())
+
+	pipelineName := tu.UniqueString("MalformedPipeline")
+
+	_, err := c.PpsAPIClient.CreatePipeline(c.Ctx(), &pps.CreatePipelineRequest{
+		Pipeline:  client.NewPipeline(pipelineName),
+		Transform: &pps.Transform{Cmd: []string{"bash"}, Stdin: []string{"echo foo > /pfs/out/file"}},
+		Input:     &pps.Input{Cron: &pps.CronInput{Spec: "@every 10s"}},
+	})
+	require.NoError(t, err)
+
+	// TODO: the worker fails to symlink active data directory - because there is no name?
+}
+
 func getObjectCountForRepo(t testing.TB, c *client.APIClient, repo string) int {
 	pipelineInfos, err := c.ListPipeline()
 	require.NoError(t, err)

--- a/src/server/pfs/server/driver.go
+++ b/src/server/pfs/server/driver.go
@@ -1851,6 +1851,9 @@ func (d *driver) resolveCommit(stm col.STM, userCommit *pfs.Commit) (*pfs.Commit
 	if userCommit == nil {
 		return nil, errors.Errorf("cannot resolve nil commit")
 	}
+	if userCommit.Repo == nil {
+		return nil, errors.Errorf("cannot resolve commit with no repo")
+	}
 	if userCommit.ID == "" {
 		return nil, errors.Errorf("cannot resolve commit with no ID or branch")
 	}

--- a/src/server/pps/server/api_server.go
+++ b/src/server/pps/server/api_server.go
@@ -261,6 +261,9 @@ func (a *apiServer) validateInput(pachClient *client.APIClient, pipelineName str
 					return errors.Errorf("multiple input types set")
 				}
 				set = true
+				if len(input.Cron.Name) == 0 {
+					return errors.Errorf("input must specify a name")
+				}
 				if _, err := cron.ParseStandard(input.Cron.Spec); err != nil {
 					return errors.Wrapf(err, "error parsing cron-spec")
 				}

--- a/src/server/pps/server/api_server.go
+++ b/src/server/pps/server/api_server.go
@@ -2229,9 +2229,9 @@ func (a *apiServer) CreatePipeline(ctx context.Context, request *pps.CreatePipel
 		var commit *pfs.Commit
 		if request.SpecCommit != nil {
 			// Make sure that the spec commit actually exists
-			commitInfo, err := pachClient.InspectCommit(request.SpecCommit.Repo.Name, request.SpecCommit.ID)
+			commitInfo, err := pachClient.PfsAPIClient.InspectCommit(pachClient.Ctx(), &pfs.InspectCommitRequest{Commit: request.SpecCommit})
 			if err != nil {
-				return nil, errors.Wrapf(err, "error inspecting commit: \"%s@%s\"", request.SpecCommit.Repo.Name, request.SpecCommit.ID)
+				return nil, errors.Wrap(err, "error inspecting spec commit")
 			}
 			// It does, so we use that as the spec commit, rather than making a new one
 			commit = commitInfo.Commit

--- a/src/server/worker/driver/driver_unix.go
+++ b/src/server/worker/driver/driver_unix.go
@@ -111,6 +111,9 @@ func (d *driver) linkData(inputs []*common.Input, dir string) error {
 		if input.S3 {
 			continue // S3 data is not downloaded
 		}
+		if input.Name == "" {
+			return errors.New("input does not have a name")
+		}
 		src := filepath.Join(dir, input.Name)
 		dst := filepath.Join(d.InputDir(), input.Name)
 		if err := os.Symlink(src, dst); err != nil {

